test/e2e/util.go

```diff
@@ -182,3 +186,121 @@ func randomSuffix() string {
+	}
+}
+
+type validatorFn func(c *client.Client, podId string) (interface{}, error)
+
+/**
+* Abstracted version of the original validateController function.
+*
+* Takes a "containername": this is grepped for.
+* Takes a "testname":  which gets bubbled up to the logging/failure messages if errors happen.
+* Takes a "validator" function: This function is given a podID and a client, and it can do some specific validations that way.
+ */
+func validateController(c *client.Client, image string, replicas int, containername string, testname string, validator validatorFn) {
+
+	getPodsTemplate := "--template={{range.items}}{{.id}} {{end}}"
+
+	// NB: kubectl adds the "exists" function to the standard template functions.
+	// This lets us check to see if the "running" entry exists for each of the containers
+	// we care about. Exists will never return an error and it's safe to check a chain of
+	// things, any one of which may not exist. In the below template, all of info,
+	// containername, and running might be nil, so the normal index function isn't very
+	// helpful.
+	// This template is unit-tested in kubectl, so if you change it, update the unit test.
+	//
+	// You can read about the syntax here: http://golang.org/pkg/text/template/
+	containername = "update-demo"
+	getContainerStateTemplate := fmt.Sprintf(`--template={{and (exists . "currentState" "info" "%s" "state" "running")}}`, containername)
+
+	getImageTemplate := fmt.Sprintf(`--template={{(index .currentState.info "%s").image}}`, containername)
+
+	By(fmt.Sprintf("waiting for all containers in %s pods to come up.", testname)) //testname should be selector
+	for start := time.Now(); time.Since(start) < podStartTimeout; time.Sleep(5 * time.Second) {
+		getPodsOutput := runKubectl("get", "pods", "-o", "template", getPodsTemplate, "-l", testname)
+		pods := strings.Fields(getPodsOutput)
+		if numPods := len(pods); numPods != replicas {
+			By(fmt.Sprintf("Replicas for %s: expected=%d actual=%d", testname, replicas, numPods))
+			continue
+		}
+		var runningPods []string
+		for _, podID := range pods {
+			running := runKubectl("get", "pods", podID, "-o", "template", getContainerStateTemplate)
+			if running == "false" {
+				Logf("%s is created but not running", podID)
+				continue
+			}
+
+			currentImage := runKubectl("get", "pods", podID, "-o", "template", getImageTemplate)
+			if currentImage != image {
+				Logf("%s is created but running wrong image; expected: %s, actual: %s", podID, image, currentImage)
+				continue
+			}
+
+			//call the generic validator function here
+			// this might validate for example, that (1) getting a url works and (2) url is serving correct content.
+			_, err := validator(c, podID)
+			if err != nil {
+				Logf("%s is running right image but validator function failed: %v", podID, err)
+				continue
+			}
+
+			Logf("%s is verified up and running", podID)
+			runningPods = append(runningPods, podID)
+		}
+		//If we reach here, then all our checks passed
```

Comment:
Comments should be full sentences (https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences) and you should leave a space between the // and the beginning of the comment.